dev/core#3965 APIv4 - Add Mailing entities for SearchKit

colemanw opened the pull request:
Overview
----------------------------------------
Exposes a bunch of MailingEvent APIs to SearchKit, and cleans up the labels in the metadata to make it a bit more user-friendly.
See https://lab.civicrm.org/dev/core/-/issues/3965

Before
---------------
SearchKit can't produce mailing event reports e.g. "Bounce Report" or "Click-Through Report"

After
------------
Now it can, although it takes a lot of joins to do so:

![image](https://user-images.githubusercontent.com/2874912/204115224-deef4a73-80fa-4a31-a86e-1e2c7dde2076.png)

Comments
------------
As noted above, producing a mailing event report takes a lot of joins. At minimum:

`MailingEventBounce -> MailingEventQueue > MailingJob -> Mailing`

I played around with the idea of making either `MailingEventQueue` or `MailingJob` into Bridge entities, however neither one of them is a serviceable bridge, at least not the way a bridge entitiy is currently defined in APIv4, where it needs to have 2 FK columns within the bridge table that join 2 other entities together. That's not the way these joins work; in the above progression each entity has a single FK column that joins to the next entity down the line (following the direction of the arrows). With a Bridge entity you'd need at least one of the arrows reversed so it's pointing *away* from the bridge in either direction. 

So I don't know how to make these searches more user-friendly to build, but I do know that Olly and Chris at NorthBridgeDigital have been working on creating pre-packaged searches for CiviCRM that would give a template to follow so these joins didn't have to be selected from scratch.

colemanw committed:
Tag class aliases as @deprecated

colemanw committed:
dev/core#3965 Add Mailng Events to APIv4

eileenmcnaughton commented:
api\v4\Entity\ConformanceTest::testConformance with data set "MailingEventBounce" ('MailingEventBounce')
Undefined array key "job_id"

/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/CRM/Mailing/Event/BAO/MailingEventBounce.php:27
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/Api4/Generic/Traits/DAOActionTrait.php:163
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/Api4/Generic/Traits/DAOActionTrait.php:134
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/Api4/Generic/DAOCreateAction.php:35
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/Api4/Provider/ActionObjectProvider.php:72
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/API/Kernel.php:149
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/Civi/Api4/Generic/AbstractAction.php:250
/home/jenkins/bknix-dfl/build/core-25059-381f3/web/sites/all/modules/civicrm/tests/phpunit/api/v4/Entity/ConformanceTest.php:257
/home/jenkins/bknix-dfl/build/core-25059-

colemanw committed:
MailingEventBounce - Rename `create` function to `recordBounce`

Because it's not a standard CRUD create function it was messing up the APIv4 syntaxConformanceTest
Best rename it for clarity.

eileenmcnaughton commented:
We might just about be able to close https://lab.civicrm.org/dev/core/-/issues/2486 now

larssandergreen commented:
Thanks @colemanw, glad to see this get added.

Maybe a bigger question, but is there any thought of adding more abstract entities to API4? As you note, it's pretty complex to get mailing data. Templates may help somewhat, but would be somewhat limiting. For example, I often want to do something like Contacts with Contributions with Mailing Clicks from a certain set of mailings — and that looks like it will be a real pain to do. An abstract entity that combined allowed this to be done more simply seems like it would be very helpful (as discussed in [the issue](https://lab.civicrm.org/dev/core/-/issues/3965)), but I'm not sure what the practical aspects of that would be.

Is that something that we could envision as part of the API in the future? Or is there another way to make this more user friendly?
